truenas_api_client/__init__.py

```diff
@@ -314,85 +313,76 @@ def _send(self, data):
+                                for event in self._event_callbacks['*']:
+                                    self._run_callback(event, [params['msg'].upper()], params)
+                            if params['collection'] in self._event_callbacks:
+                                for event in self._event_callbacks[params['collection']]:
+                                    self._run_callback(event, [params['msg'].upper()], params)
+                    case 'notify_unsubscribed':
+                        if params['collection'] in self._event_callbacks:
+                            for event in self._event_callbacks[params['collection']]:
+                                if 'error' in params:
+                                    event['error'] = params['error']['reason'] or params['error']
+                                event['event'].set()
+                    case _:
+                        logger.error('Received unknown notification %r', message['method'])
+            elif 'id' in message:
+                if self._set_options_call and message['id'] == self._set_options_call.id:
                     if 'error' in message:
-                        event['error'] = message['error']['reason'] or message['error']['error']
-                    event['ready'].set()
-                    event['event'].set()
+                        try:
+                            self._parse_error(message['error'], self._set_options_call)
+                        except Exception:
+                            logger.error('Unhandled exception in Client._parse_error', exc_info=True)
+                        else:
+                            logger.error('Error setting client options: %r', self._set_options_call.error)
+                    self._connected.set()
+                elif call := self._calls.get(message['id']):
+                    if 'result' in message:
+                        call.result = message['result']
+                    if 'error' in message:
+                        try:
+                            self._parse_error(message['error'], call)
+                        except Exception:
+                            logger.error('Unhandled exception in Client._parse_error', exc_info=True)
+                    call.returned.set()
+                    self._unregister_call(call)
+                else:
+                    logger.error('Received a response for non-registered method call %r', message['id'])
+            else:
+                logger.error('Received unknown message %r', message)
+        except Exception:
+            logger.error('Unhandled exception in Client._recv', exc_info=True)
+
+    def _parse_error(self, error: dict, call: Call):
+        code = JSONRPCError(error['code'])
+        if self._py_exceptions and code in [JSONRPCError.INVALID_PARAMS, JSONRPCError.TRUENAS_CALL_ERROR]:
+            data = error['data']
+            call.error = ClientException(data['reason'], data['error'], data['trace'], data['extra'])
```

Comment:
`error['data']` could be a string here if the error was sent as a result of catching `jsonschema.ValidationError` in `RpcWebSocketHandler.process_message`. Either [this line](https://github.com/truenas/middleware/pull/13979/files#diff-e4af733f1ec823e6ed116943c1f9dea119a79baaf38d75b13a6b137c6c22754dR251) needs to send a `dict` or `_parse_error` needs to handle when `error['data']` is a string.

Reply:
The caller was incorrect.